sql/stats: unset histogram buckets proto after decoding

yuzefovich opened the pull request:
**sql/stats,backup: adjust table stats handling in backup**

This commit introduces a new method to return the protos of table
statistics directly from the DB (i.e. it by-passes the table stats cache)
and uses this method in the backup code. The main motivation is that we
want to nil out the histogram buckets in the proto that we store in the
table stats cache since we don't need them outside of the backup.

The new method hardens the logic of excluding merged and forecast stats
from the backup because we simply no longer include these stats into
what backups includes (because they are generated on demand rather than
read from the system table). It additionally ignores whether "stats
usage is allowed" on a particular table since it seems beneficial to
backup stats for all requested tables, and then once stats are restored
we can decide whether to actually use them.

Notable change for the new `GetTableStatsProtosFromDB` method is that we
no longer hydrate the column type nor decode upper bounds since this isn't
needed for the backup use case.

**stats: unset histogram buckets proto after decoding**

I think it should be safe to unset `HistogramData.Buckets` proto after
having decoded the histogram buckets since we only use `cat.HistogramBucket`
with decoded datums going forward (the only exception was the backup use
case, but it was adjusted in the previous commit to have a separate code
path). I noticed the proto was showing up noticeably in the inuse-space
heap profiles on a customer cluster.

It also adds a comment for why we're use `*bool` as `forecast` argument
in `getTableStatsFromCache`, as opposed to just `bool`.

Epic: None
Release note: None

yuzefovich committed:
stats: unset histogram buckets proto after decoding

I think it should be safe to unset `HistogramData.Buckets` proto after
having decoded the histogram buckets since we only use `cat.HistogramBucket`
with decoded datums going forward (the only exception was the backup use
case, but it was adjusted in the previous commit to have a separate code
path). I noticed the proto was showing up noticeably in the inuse-space
heap profiles on a customer cluster.

It also adds a comment for why we're use `*bool` as `forecast` argument
in `getTableStatsFromCache`, as opposed to just `bool`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142963)
<!-- Reviewable:end -->

yuzefovich committed:
sql/stats,backup: adjust table stats handling in backup

This commit introduces a new method to return the protos of table
statistics directly from the DB (i.e. it by-passes the table stats cache)
and uses this method in the backup code. The main motivation is that we
want to nil out the histogram buckets in the proto that we store in the
table stats cache since we don't need them outside of the backup.

The new method hardens the logic of excluding merged and forecast stats
from the backup because we simply no longer include these stats into
what backups includes (because they are generated on demand rather than
read from the system table). It additionally ignores whether "stats
usage is allowed" on a particular table since it seems beneficial to
backup stats for all requested tables, and then once stats are restored
we can decide whether to actually use them.

Notable change for the new `GetTableStatsProtosFromDB` method is that we
no longer hydrate the column type nor decode upper bounds since this isn't
needed for the backup use case.

Release note: None

yuzefovich commented:
I tried to reproduce the same heap profile that I saw when looking at [this](https://github.com/cockroachlabs/support/issues/3212) ticket but didn't succeed. In other words, I wasn't able manually construct a scenario where the memory usage of this proto would show up noticeably in the heap profile. Still, it seems like an improvement overall to me, so let's proceed with the change.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 4 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-:-OLdsNONA-DGlulbEiQ_:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

___
*[`-- commits` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLdbm6t9mZFZyd9Hsl5:-OLdbm6uBS2MfrE5n3tY:b-akxezs):*
nit: you used a different name, I think -- `getTableStatsProtosFromDB`
___
*[`-- commits` line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLdbxr1D1yGpdNQyKUC:-OLdbxr1D1yGpdNQyKUD:b-7m534n):*
Is there no chance that any historical backup includes them?
___
*[`pkg/sql/stats/stats_cache.go` line 937 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLdcc4N0eJPpQ6ypWgi:-OLdcc4N0eJPpQ6ypWgj:b-k8uhm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5513e71ba94c8c12e6914af1f5a4fa0654f25a9/pkg/sql/stats/stats_cache.go#L937)):*
> ```Go
> // for the given table ID, in "un-decoded" protobuf form.
> //
> // It ignores any statistics that cannot be decoded (e.g. because a user-defined
> ```

nit: because -> because of (ditto above)


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
Nice! This is a great idea.

> Reviewed 4 of 4 files at r1, 4 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-:-OLexG3u5kVWgJmgo6ca:bclmtrk)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/stats/stats_cache.go` line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLetbtY0aiVZ7j7zplh:-OLetbtY0aiVZ7j7zpli:b-mm8moy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5513e71ba94c8c12e6914af1f5a4fa0654f25a9/pkg/sql/stats/stats_cache.go#L253)):*
> ```Go
> // The function returns an error if we could not query the system table. It
> // silently ignores any statistics that can't be decoded (e.g. because
> // user-defined types don't exit).
> ```

typo: "exist"
___
*[`pkg/sql/stats/stats_cache.go` line 262 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLeuuY2Eu-G41RV0ABh:-OLeuuY2Eu-G41RV0ABi:b-yzu6oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5513e71ba94c8c12e6914af1f5a4fa0654f25a9/pkg/sql/stats/stats_cache.go#L262)):*
> ```Go
> 	settings *cluster.Settings,
> ) (statsProtos []*TableStatisticProto, err error) {
> 	if !statsUsageAllowed(table, settings) {
> ```

Very minor point, but I wonder if we should elide this check? It seems like we should back up all table stats in the system, regardless of whether `sql.stats.system_tables.enabled` is set or not. And then maybe we wouldn't need to pass in settings?
___
*[`pkg/sql/stats/stats_cache.go` line 943 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLeu-j46v4F-YsxI6Uo:-OLeu-j46v4F-YsxI6Up:bjzt6ga) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5513e71ba94c8c12e6914af1f5a4fa0654f25a9/pkg/sql/stats/stats_cache.go#L943)):*
> ```Go
> ) (statsProtos []*TableStatisticProto, err error) {
> 	it, err := executor.QueryIteratorEx(
> 		ctx, "get-table-statistics", nil /* txn */, sessiondata.NodeUserSessionDataOverride, getTableStatisticsStmt, tableID,
> ```

I know it's the same query, but it might be worth giving this execution a slightly different name for debugging purposes.
___
*[`pkg/backup/backup_job.go` line 503 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLevdIv7AqjEkXVVfsm:-OLevdIv7AqjEkXVVfsn:b-575boh) ([raw file](https://github.com/cockroachdb/cockroach/blob/96298a45be3134b4c3ed9e884d9980a9721db2e4/pkg/backup/backup_job.go#L503)):*
> ```Go
> 	// stats on demand and do not need to be backed up or
> 	// restored.
> 	return stat.Name != jobspb.ForecastStatsName && stat.Name != jobspb.MergedStatsName
> ```

I agree that this check shouldn't catch anything, but I think we should keep it. It doesn't hurt anything, and it also guards against any hypothetical future bugs accidentally including forecasts or merged stats in a backup that is then restored by an earlier version.
___
*[`pkg/sql/stats/stats_cache.go` line 359 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLeq7aw0r6b_svB3wbE:-OLeq7aw0r6b_svB3wbF:b-5y8xhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/96298a45be3134b4c3ed9e884d9980a9721db2e4/pkg/sql/stats/stats_cache.go#L359)):*
> ```Go
> func (e *cacheEntry) isStale(forecast *bool, udtCols []catalog.Column) bool {
> 	// Check whether forecast settings have changed.
> 	if forecast != nil && e.forecast != *forecast {
> ```

The purpose of the *bool is to prevent tripping the `(*cacheEntry).isStale` check when we call `getTableStatsFromCache` from `stats.(*Refresher).maybeRefreshStats`. The refresher needs to check when the most recent automatic full collection was, but it doesn't care whether there is a forecast or not in the stats (it will ignore the forecast if there is one). Using nil is a way to skip part of the isStale check rather than unnecessarily invalidate the stats cache entry.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the reviews!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-:-OLfkVtm16MiLArMH7Rc:b-hlc8bm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rytaft)

___
*[`-- commits` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLdbm6t9mZFZyd9Hsl5:-OLfgtPfEMHytDGh8WXE:b-44rl1z):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: you used a different name, I think -- `getTableStatsProtosFromDB`
</blockquote></details>

Yeah, I used a different name for a new function, but here I'm referring to an old existing function.
___
*[`-- commits` line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLdbxr1D1yGpdNQyKUC:-OLfh1zQ-JQ__DpRYdGR:b-bdu42e):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is there no chance that any historical backup includes them?
</blockquote></details>

We only need to be able to restore backups taken on 24.3+ into 25.2 version [thread](https://cockroachlabs.slack.com/archives/C03JCUUSCD6/p1742243085176139), and since we've excluded merged and forecast stats since about 23.1, so no eligible backups should have them.

Still, given Michael's comment, I brought the check back.
___
*[`pkg/backup/backup_job.go` line 503 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLevdIv7AqjEkXVVfsm:-OLfhzwm7Ww_cu_skAF1:b-5dxjbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/96298a45be3134b4c3ed9e884d9980a9721db2e4/pkg/backup/backup_job.go#L503)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I agree that this check shouldn't catch anything, but I think we should keep it. It doesn't hurt anything, and it also guards against any hypothetical future bugs accidentally including forecasts or merged stats in a backup that is then restored by an earlier version.
</blockquote></details>

I don't think we allow restoring newer backups into older versions, so I doubt this check will do anything ever (unless we modify the behavior of `GetTableStatsProtosFromDB`), but I agree that having this check doesn't have any cost, so I brought it back.
___
*[`pkg/sql/stats/stats_cache.go` line 359 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLeq7aw0r6b_svB3wbE:-OLffJMw-ScOht1-5HVG:b-xzmkph) ([raw file](https://github.com/cockroachdb/cockroach/blob/96298a45be3134b4c3ed9e884d9980a9721db2e4/pkg/sql/stats/stats_cache.go#L359)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

The purpose of the *bool is to prevent tripping the `(*cacheEntry).isStale` check when we call `getTableStatsFromCache` from `stats.(*Refresher).maybeRefreshStats`. The refresher needs to check when the most recent automatic full collection was, but it doesn't care whether there is a forecast or not in the stats (it will ignore the forecast if there is one). Using nil is a way to skip part of the isStale check rather than unnecessarily invalidate the stats cache entry.
</blockquote></details>

Ack, thanks for the context. I included this as a comment to prevent confusion in the future.
___
*[`pkg/sql/stats/stats_cache.go` line 262 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLeuuY2Eu-G41RV0ABh:-OLfjb2h-HYNYA46h4CR:b-v0w5m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5513e71ba94c8c12e6914af1f5a4fa0654f25a9/pkg/sql/stats/stats_cache.go#L262)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Very minor point, but I wonder if we should elide this check? It seems like we should back up all table stats in the system, regardless of whether `sql.stats.system_tables.enabled` is set or not. And then maybe we wouldn't need to pass in settings?
</blockquote></details>

I agree, this seems like a better decision. Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice! Thank you for doing this!

> Reviewed 7 of 7 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-:-OLfoSF509ij1BnXMBkP:b-808s1r)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[`pkg/sql/stats/stats_cache.go` line 359 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142963#-OLeq7aw0r6b_svB3wbE:-OLfoqysCqCfPRbd7NHD:b-dzh0cj) ([raw file](https://github.com/cockroachdb/cockroach/blob/96298a45be3134b4c3ed9e884d9980a9721db2e4/pkg/sql/stats/stats_cache.go#L359)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ack, thanks for the context. I included this as a comment to prevent confusion in the future.
</blockquote></details>

Thanks! Sorry, I should have left a comment originally.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
